router/sleeve.go

```diff
@@ -0,0 +1,1117 @@
+
+	if fwd.heartbeatInterval != 0 {
+		// already confirmed
+		return nil
+	}
+
+	// heartbeatInterval flags that we want to send heartbeats,
+	// even if we don't do sendHeartbeat() yet due to lacking the
+	// remote address.
+	fwd.heartbeatInterval = time.Duration(FastHeartbeat)
+	if fwd.remoteAddr != nil {
+		if err := fwd.sendHeartbeat(); err != nil {
+			return err
+		}
+	}
+
+	fwd.heartbeatTimeout = time.NewTimer(HeartbeatTimeout)
+	return nil
+}
+
+func (fwd *sleeveForwarder) sendHeartbeat() error {
+	log.Debug(fwd.logPrefix(), "sendHeartbeat")
+
+	// Prime the timer for the next heartbeat.  We don't use a
+	// ticker because the interval is not constant.
+	fwd.heartbeatTimer = setTimer(fwd.heartbeatTimer, fwd.heartbeatInterval)
+
+	buf := make([]byte, EthernetOverhead+8)
+	binary.BigEndian.PutUint64(buf[EthernetOverhead:], fwd.connUID)
+	return fwd.sendSpecial(fwd.crypto.EncDF, fwd.senderDF, buf)
+}
+
+func (fwd *sleeveForwarder) handleHeartbeat(special specialFrame) error {
+	uid := binary.BigEndian.Uint64(special.frame[EthernetOverhead:])
+	if uid != fwd.connUID {
+		return nil
+	}
+
+	log.Debug(fwd.logPrefix(), "handleHeartbeat")
+
+	if fwd.remoteAddr == nil {
+		fwd.setRemoteAddr(special.sender)
+		if fwd.heartbeatInterval != time.Duration(0) {
+			if err := fwd.sendHeartbeat(); err != nil {
+				return err
+			}
+		}
+	} else if !udpAddrsEqual(fwd.remoteAddr, special.sender) {
+		log.Println(fwd.logPrefix(),
+			"Peer UDP address changed to", special.sender)
+		fwd.setRemoteAddr(special.sender)
+	}
+
+	if !fwd.ackedHeartbeat {
+		fwd.ackedHeartbeat = true
+		if err := fwd.sendControlMsg([]byte{HeartbeatAck}); err != nil {
+			return err
+		}
+	}
+
+	fwd.heartbeatTimeout.Reset(HeartbeatTimeout)
+	return nil
+}
+
+func (fwd *sleeveForwarder) setRemoteAddr(addr *net.UDPAddr) {
+	// Although we don't need to lock when reading remoteAddr,
+	// because this thread is the only one that modifies
+	// remoteAddr, we do need to lock when writing it, because
+	// memory models.
+	fwd.lock.Lock()
+	fwd.remoteAddr = addr
+	fwd.lock.Unlock()
+}
+
+func (fwd *sleeveForwarder) handleHeartbeatAck() error {
+	// The connection is nowregarded as established
+	fwd.notifyEstablished()
+
+	if fwd.heartbeatInterval != SlowHeartbeat {
+		fwd.heartbeatInterval = SlowHeartbeat
+		if fwd.heartbeatTimer != nil {
+			fwd.heartbeatTimer.Reset(fwd.heartbeatInterval)
+		}
+	}
+
+	fwd.fragTestTicker = time.NewTicker(FragTestInterval)
+	if err := fwd.sendFragTest(); err != nil {
+		return err
+	}
+
+	// Send a large frame down the DF channel in order to prompt
+	// PMTU discovery to start.
+	return fwd.sendSpecial(fwd.crypto.EncDF, fwd.senderDF,
+		make([]byte, PMTUDiscoverySize))
+}
+
+func (fwd *sleeveForwarder) notifyEstablished() {
+	fwd.lock.RLock()
+	defer fwd.lock.RUnlock()
+	if fwd.listener != nil {
+		fwd.listener.Established()
+	}
+}
+
+func (fwd *sleeveForwarder) sendFragTest() error {
+	log.Debug(fwd.logPrefix(), "sendFragTest")
+	fwd.stackFrag = false
```

Comment:
> I appreciate that the previous implementation did the same thing, but I find this a bit odd - we're setting stackFrag to false periodically (with concomitant effects in Send) even when stack fragmentation is continuously available and working. The only benefit from doing this that I can think of is avoiding an additional timer to notice a lack of FragTestAck - is that your assessment too?

As you say, it is doing what the old implementation did.

`stackFrag == false` should only have implications for performance, not behaviour, and only for packets that are larger than the MTU, which tend to be rare in practice.  And other normal circumstances, it will be set back to true promptly.  It doesn't seem like a major concern.

> Also shouldn't the modification of fwd.stackFrag here and in handleFragTestAck be protected by a write lock as Send (which obtains a read lock before accessing stackFrag) is called from a different goroutine?

Although the golang memory model is entirely undocumented AFAICT, and I have not been able to find an authoritative pronouncement on the topic, it seems safe to assume that reads and writes of bools are atomic. They are compiled to single bytes, not bit-packed (the ability to take pointers to them makes this almost inevitable). So modifying it without holding a lock is safe.